File: Data Structures & Algorithms/eating-bananas/submission-0.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # take from range 1 to max(piles) and binary search each possible k through O(n) for each k
        l = 1
        r = max(piles)
        res = r
        while l <= r:
            m = (l + r) // 2
            hours = 0
            for p in piles:
                hours += math.ceil(p / m)
            
            if hours <= h:
                res = min(res, m)
                r = m - 1
            else:
                l = m + 1
        return res

File: Data Structures & Algorithms/binary-search/submission-0.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if (nums.size() == 0) return -1;
        int m = nums.size() / 2;
        if (target == nums[m]) {
            return m;
        } else if (target < nums[m]) {
            vector<int> v(nums.begin(), nums.begin() + m);
            return search(v, target);
        } else {
            vector<int> v(nums.begin() + m + 1, nums.end());
            int a = search(v, target);
            if (a != -1) {
                return a + m + 1;
            }
        }
        return -1;
    }
};
